File: sample/OpenCvSharp.Samples/Samples/FaceDetect.cs
//#define CHECK_MEMORY_LEAK

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace OpenCvSharp.Test
{
    /// <summary>
    /// 顔の検出
    /// </summary>
    /// <remarks>http://opencv.jp/sample/object_detection.html#face_detection</remarks>
    class FaceDetect
    {
        public FaceDetect()
        {
            CheckMemoryLeak();

            // CvHaarClassifierCascade, cvHaarDetectObjects
            // 顔を検出するためにHaar分類器のカスケードを用いる

            CvColor[] colors = new CvColor[]{
                new CvColor(0,0,255),
                new CvColor(0,128,255),
                new CvColor(0,255,255),
                new CvColor(0,255,0),
                new CvColor(255,128,0),
                new CvColor(255,255,0),
                new CvColor(255,0,0),
                new CvColor(255,0,255),
            };

            const double Scale = 1.14;
            const double ScaleFactor = 1.0850;
            const int MinNeighbors = 2;

            using (IplImage img = new IplImage(Const.ImageYalta, LoadMode.Color))
            using (IplImage smallImg = new IplImage(new CvSize(Cv.Round(img.Width / Scale), Cv.Round(img.Height / Scale)), BitDepth.U8, 1))
            {
                // 顔検出用の画像の生成
                using (IplImage gray = new IplImage(img.Size, BitDepth.U8, 1))
                {
                    Cv.CvtColor(img, gray, ColorConversion.BgrToGray);
                    Cv.Resize(gray, smallImg, Interpolation.Linear);
                    Cv.EqualizeHist(smallImg, smallImg);
                }

                //using (CvHaarClassifierCascade cascade = Cv.Load<CvHaarClassifierCascade>(Const.XmlHaarcascade))  // どっちでも可
                using (CvHaarClassifierCascade cascade = CvHaarClassifierCascade.FromFile(Const.XmlHaarcascade))    // 
                using (CvMemStorage storage = new CvMemStorage())
                {
                    storage.Clear();

                    // 顔の検出
                    Stopwatch watch = Stopwatch.StartNew();
                    CvSeq<CvAvgComp> faces = Cv.HaarDetectObjects(smallImg, cascade, storage, ScaleFactor, MinNeighbors, 0, new CvSize(30, 30));
                    watch.Stop();
                    Console.WriteLine("detection time = {0}ms\n", watch.ElapsedMilliseconds);

                    // 検出した箇所にまるをつける
                    for (int i = 0; i < faces.Total; i++)
                    {
                        CvRect r = faces[i].Value.Rect;
                        CvPoint center = new CvPoint
                        {
                            X = Cv.Round((r.X + r.Width * 0.5) * Scale),
                            Y = Cv.Round((r.Y + r.Height * 0.5) * Scale)
                        };
                        int radius = Cv.Round((r.Width + r.Height) * 0.25 * Scale);
                        img.Circle(center, radius, colors[i % 8], 3, LineType.AntiAlias, 0);
                    }
                }

                // ウィンドウに表示
                CvWindow.ShowImages(img);
            }
        }

        [Conditional("CHECK_MEMORY_LEAK")]
        private void CheckMemoryLeak()
        {            
            while(true)
            {
                Console.WriteLine("{0} Kbytes", GC.GetTotalMemory(false) / 1024);
                ConsoleKeyInfo keyInfo = Console.ReadKey();
                switch (keyInfo.Key)
                {
                    case ConsoleKey.Enter:
                        for (int i = 0; i < 128; i++)
                        {
                            using (CvHaarClassifierCascade cascade = CvHaarClassifierCascade.FromFile(Const.XmlHaarcascade))
                            {                                
                            }
                            //using (CvMat mat = new CvMat(1, 1024, MatrixType.U8C1))
                            //{                                
                            //}
                        }
                        GC.Collect();
                        break;
                    case ConsoleKey.Escape:
                        goto END_LOOP;
                }
                
            }
        END_LOOP: ;
        }

    }
}

File: sample_VB_Neoxeo/OpenCvSharp.Samples/Samples/FaceDetect.vb
'#define CHECK_MEMORY_LEAK

Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Linq
Imports System.Text
Imports OpenCvSharp

Namespace OpenCvSharpSamples
	''' <summary>
	''' 顔の検出
	''' </summary>
	''' <remarks>http://opencv.jp/sample/object_detection.html#face_detection</remarks>
    Friend Module FaceDetect
        Public Sub Start()
            CheckMemoryLeak()

            ' CvHaarClassifierCascade, cvHaarDetectObjects
            ' 顔を検出するためにHaar分類器のカスケードを用いる

            Dim colors() As CvColor = { _
                New CvColor(0, 0, 255), _
                New CvColor(0, 128, 255), _
                New CvColor(0, 255, 255), _
                New CvColor(0, 255, 0), _
                New CvColor(255, 128, 0), _
                New CvColor(255, 255, 0), _
                New CvColor(255, 0, 0), _
                New CvColor(255, 0, 255) _
            }

            Const Scale As Double = 1.14
            Const ScaleFactor As Double = 1.085
            Const MinNeighbors As Integer = 2

            Using img As New IplImage([Const].ImageYalta, LoadMode.Color)
                Using smallImg As New IplImage(New CvSize(Cv.Round(img.Width / Scale), Cv.Round(img.Height / Scale)), BitDepth.U8, 1)
                    ' 顔検出用の画像の生成
                    Using gray As New IplImage(img.Size, BitDepth.U8, 1)
                        Cv.CvtColor(img, gray, ColorConversion.BgrToGray)
                        Cv.Resize(gray, smallImg, Interpolation.Linear)
                        Cv.EqualizeHist(smallImg, smallImg)
                    End Using

                    'using (CvHaarClassifierCascade cascade = Cv.Load<CvHaarClassifierCascade>(Const.XmlHaarcascade))  // どっちでも可
                    Using cascade As CvHaarClassifierCascade = CvHaarClassifierCascade.FromFile([Const].XmlHaarcascade)
                        Using storage As New CvMemStorage()
                            storage.Clear()

                            ' 顔の検出
                            Dim watch As Stopwatch = Stopwatch.StartNew()
                            Dim faces As CvSeq(Of CvAvgComp) = Cv.HaarDetectObjects(smallImg, cascade, storage, ScaleFactor, MinNeighbors, 0, New CvSize(30, 30))
                            watch.Stop()
                            Form1.TextBox1.AppendText(Environment.NewLine & String.Format("detection time = {0}ms" & vbLf, watch.ElapsedMilliseconds))

                            ' 検出した箇所にまるをつける
                            For i As Integer = 0 To faces.Total - 1
                                Dim r As CvRect = faces(i).Value.Rect
                                Dim center As CvPoint = New CvPoint With {.X = Cv.Round((r.X + r.Width * 0.5) * Scale), .Y = Cv.Round((r.Y + r.Height * 0.5) * Scale)}
                                Dim radius As Integer = Cv.Round((r.Width + r.Height) * 0.25 * Scale)
                                img.Circle(center, radius, colors(i Mod 8), 3, LineType.AntiAlias, 0)
                            Next i
                        End Using
                    End Using

                    ' ウィンドウに表示
                    CvWindow.ShowImages(img)
                End Using
            End Using
        End Sub

        <Conditional("CHECK_MEMORY_LEAK")> _
        Private Sub CheckMemoryLeak()
            Do
                Form1.TextBox1.AppendText(Environment.NewLine & String.Format("{0} Kbytes", GC.GetTotalMemory(False) \ 1024))
                Console.WriteLine("{0} Kbytes", GC.GetTotalMemory(False) \ 1024)
                Dim keyInfo As ConsoleKeyInfo = Console.ReadKey()
                Select Case keyInfo.Key
                    Case ConsoleKey.Enter
                        For i As Integer = 0 To 127
                            Using cascade As CvHaarClassifierCascade = CvHaarClassifierCascade.FromFile([Const].XmlHaarcascade)
                            End Using
                            'using (CvMat mat = new CvMat(1, 1024, MatrixType.U8C1))
                            '{                                
                            '}
                        Next i
                        GC.Collect()
                    Case ConsoleKey.Escape
                        GoTo END_LOOP
                End Select

            Loop
END_LOOP:

        End Sub

    End Module
End Namespace
